### PR TITLE
Improve support for joda datetime to java datetime in Painless

### PR DESCRIPTION
**Pull request by jdconrad**
Within Painless, this change augments `TemporalAccessor` with the method `getMillis` to make milliseconds easier to access directly from a datetime within a document.

This change also adds several methods from Joda datetimes augmented onto `ZonedDateTime` that throw an `UnsupportedOperationException` with a suggestion of how to implement that Joda method using `ZonedDateTime` instead. These methods include the following:
* `getCenturyOfEra`
* `getEra`
* `getHourOfDay`
* `getMillisOfDay`
* `getMillisOfSecond`
* `getMinuteOfDay`
* `getMinuteOfHour`
* `getMonthOfYear`
* `getSecondOfDay`
* `getSecondOfMinute`
* `getWeekOfWeekyear`
* `getWeekyear`
* `getYearOfCentury`
* `getYearOfEra`

### CODE REVIEW

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/packaging-tests-unix-sample

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java [hunk 1]
```diff
@@ -708,11 +708,106 @@ public static Matcher matcher(Pattern receiver, int limitFactor, CharSequence in
...
+    }
+
+    public static long getMillis(TemporalAccessor receiver) {
+        return toEpochMilli(receiver);
     }
 
     public static DayOfWeek getDayOfWeekEnum(ZonedDateTime receiver) {
         return receiver.getDayOfWeek();
     }
+
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:722]: I wonder if we should have a separate file or section these Augmentations that's simply unsupported.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/50_script_doc_values.yml [hunk 2]
```diff
@@ -1696,3 +1784,202 @@ setup:
...
+        rest_total_hits_as_int: true
+        body:
+          query: { term: { _id: 1 } }
+          script_fields:
+            field:
+              script:
+                source: "/* avoid stash */ $('date', null).getCenturyOfEra()"
+  - match: { error.failed_shards.0.reason.caused_by.type: "unsupported_operation_exception" }
+  - match: { error.failed_shards.0.reason.caused_by.reason:
+               "joda method [getCenturyOfEra] is no longer available; use [<ZonedDateTime instance>.get(ChronoField.YEAR_OF_ERA) / 100] instead" }
```
Comment (stu-elastic) [inline modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/50_script_doc_values.yml:1803]: For each of these suggestions, can you add a test that _uses_ the suggestion right the test that suggests the alternative?

**Comment by stu-elastic** [comment]
Can you reference the change that deprecated these files in the commit?

**Review by stu-elastic — APPROVED** [review]
These augments return the methods removed in https://github.com/elastic/elasticsearch/pull/78417

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:722]: Seems like a good idea. I would want to do that in as a follow up, though.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/50_script_doc_values.yml (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/50_script_doc_values.yml:1803]: Good call. Caught a pair of bugs for the week based suggestions!

**Pull request merged**